README.md
# mccga.rs
Machine-coded compact genetic algorithm in Rust

## In-short

The package implements the Machine-coded compact genetic algorithm defined in 

Satman, M. H. & Akadal, E. (2020). Machine Coded Compact Genetic Algorithms for Real Parameter Optimization Problems . Alphanumeric Journal , 8 (1) , 43-58 . DOI: 10.17093/alphanumeric.576919 [Link](https://dergipark.org.tr/en/pub/alphanumeric/issue/55603/576919)

## Example

Suppose that the objective function is to minimize 

```rust
fn f(x: &Vec<f64>) -> f64 {
    return (x[0] - 3.14159265).powf(2.0) + (x[1] - 2.71828).powf(2.0);
}
```

so the package allows to type 

```rust 
let mins: Vec<f64> = vec![-10000.0_f64; 2];
let maxs: Vec<f64> = vec![10000.0_f64; 2];

let result = mccga(f, &mins, &maxs, 0.001, 100000);
```

to minimize the objective function where result is a 2-element vector. One can test the result using

```rust
assert!(isequal(&result[0], 3.14159265, 0.001));
assert!(isequal(&result[1], 2.71828, 0.001)); 
```

## Other implementations

- Python (https://github.com/jbytecode/mccga.py)
- Julia (https://github.com/jmejia8/Metaheuristics.jl)
- Java (https://github.com/jbytecode/mccga.java)



CHANGELOG.md
# v0.1.6 (Upcoming version)


# v0.1.5 

- Add tests for hj (Hooke & Jeeves)
- Update documentation


# v0.1.4

- Seperate single iteration of mccga from the main mccga() function. 

# v0.1.3

- Add independent cga() implementation without binary representation of floats.


# v0.1.2

- New tests added
- In-place update of probability vector for speed increment
- Speed improvements
- Some undocumented parts have now docs.

# v0.1.1

- Some documentation added.

# v0.1.0

- Initial release

src/cga.rs
use rand::random;

pub fn sample(probvector: &Vec<f64>) -> Vec<u8> {
    let n = probvector.len();
    let mut newvector = vec![0_u8; n];
    for i in 0..n {
        if random::<f64>() < probvector[i] {
            newvector[i] = 1_u8;
        }
    }
    return newvector;
}

///
/// In-place update of probability vector.
/// If ith bit of the winner and loser are equal than does nothing.
/// If ith bit of the winner is one then increase the corresponding probability by a small number (mutrate).
/// If ith bit of the winner is zero then decrease the corresponding probability by a small number.
/// The function mutates the values in place.
pub fn update(probvector: &mut Vec<f64>, winner: &Vec<u8>, loser: &Vec<u8>, mutrate: f64) -> () {
    for i in 0..probvector.len() {
        if winner[i] != loser[i] {
            if winner[i] == 0 {
                probvector[i] = (probvector[i] - mutrate).max(0.0_f64);
            } else {
                probvector[i] = (probvector[i] + mutrate).min(1.0_f64);
            }
        }
    }
}

pub fn isequal(x: &f64, other: f64, mutrate: f64) -> bool {
    return (x - other).abs() <= mutrate;
}

pub fn isconverged(probvector: &Vec<f64>, mutrate: f64) -> bool {
    let n = probvector.len();
    let mut satisfied: usize = 0;
    for i in 0..n {
        let val = probvector[i];
        if isequal(&val, 0.0, mutrate) || isequal(&val, 1.0, mutrate) {
            satisfied += 1;
        } else {
            break;
        }
    }
    return satisfied == n;
}

pub fn single_iteration(probvector: &mut Vec<f64>, fcost: fn(&Vec<u8>) -> f64, mutrate: f64) -> () {
    let candidate1: Vec<u8> = sample(&probvector);
    let candidate2: Vec<u8> = sample(&probvector);

    let cost1: f64 = fcost(&candidate1);
    let cost2: f64 = fcost(&candidate2);

    let mut winner: &Vec<u8> = &candidate1;
    let mut loser: &Vec<u8> = &candidate2;

    if cost2 < cost1 {
        winner = &candidate2;
        loser = &candidate1;
    }

    update(probvector, winner, loser, mutrate);
}

pub fn cga(fcost: fn(&Vec<u8>) -> f64, mutrate: f64, bitlen: usize) -> Vec<u8> {
    let mut probvector: Vec<f64> = vec![0.5_f64; bitlen];

    while !isconverged(&probvector, mutrate) {
        single_iteration(&mut probvector, fcost, mutrate);
    }

    let result: Vec<u8> = sample(&probvector);

    return result;
}

#[cfg(test)]
mod tests {

    use super::cga;
    use super::isconverged;
    use super::isequal;
    use super::update;

    #[test]
    fn test_update() {
        let n: usize = 5;
        let mutrate: f64 = 0.1;
        let winner: Vec<u8> = vec![1_u8; n];
        let loser: Vec<u8> = vec![0_u8; n];
        let mut probvector: Vec<f64> = vec![0.5_f64; n];

        update(&mut probvector, &winner, &loser, mutrate);

        for i in 0..n {
            assert_eq!(probvector[i], 0.6_f64);
        }
    }

    #[test]
    fn test_isequal() {
        // Equal
        assert!(isequal(&10.0, 9.99, 0.01));

        // Not equal
        assert!(!isequal(&10.0, 9.9, 0.01));
    }

    #[test]
    fn test_isconverged() {
        // Converged
        let v1: Vec<f64> = vec![1.0, 0.0, 0.0, 1.0];
        assert!(isconverged(&v1, 0.0001));

        // Converged
        let v2: Vec<f64> = vec![0.9999999, 0.0, 0.0, 1.0];
        assert!(isconverged(&v2, 0.0001));

        // Converged
        let v3: Vec<f64> = vec![0.99999, 0.0, 0.0, 1.0];
        assert!(isconverged(&v3, 0.00001));

        // Not converged
        let v4: Vec<f64> = vec![0.9, 0.0, 0.0, 1.0];
        assert!(!isconverged(&v4, 0.0001));
    }

    #[test]
    fn test_cga_maximization() {
        fn cfmax(x: &Vec<u8>) -> f64 {
            let mut sum: f64 = 0.0;
            for i in 0..x.len() {
                sum += x[i] as f64;
            }
            return -sum;
        }
        let bitlen = 20;
        let cgaresult = cga(cfmax, 0.01, bitlen);
        for i in 0..bitlen {
            assert_eq!(cgaresult[i], 1_u8);
        }
    }

    #[test]
    fn test_cga_minimization() {
        fn cfmin(x: &Vec<u8>) -> f64 {
            let mut sum: f64 = 0.0;
            for i in 0..x.len() {
                sum += x[i] as f64;
            }
            return sum;
        }
        let bitlen = 20;
        let cgaresult = cga(cfmin, 0.01, bitlen);
        for i in 0..bitlen {
            assert_eq!(cgaresult[i], 0_u8);
        }
    }
}


src/mccga.rs
use crate::byteworks::{bits2floats, floatstodoubles};
use crate::cga::{isconverged, sample, update};
use crate::genprobs::generate_probability_vector;
use crate::hookejeeves::hj;

fn binarycost(fcost: fn(&Vec<f64>) -> f64, candidate: &Vec<u8>) -> f64 {
    let floats = bits2floats(candidate);
    let doubles = floatstodoubles(&floats);
    let result = fcost(&doubles);
    return result;
}

pub fn mccga(
    fcost: fn(&Vec<f64>) -> f64,
    mins: &Vec<f64>,
    maxs: &Vec<f64>,
    mutrate: f64,
    maxiter: usize,
) -> Vec<f64> {
    let mut probvect = generate_probability_vector(mins, maxs, 500000);

    let mut iter = 0;

    while iter < maxiter {
        let candidate1: Vec<u8> = sample(&probvect);
        let candidate2: Vec<u8> = sample(&probvect);

        let cost1 = binarycost(fcost, &candidate1);
        let cost2 = binarycost(fcost, &candidate2);

        let mut winner = &candidate1;
        let mut loser = &candidate2;

        if cost2 < cost1 {
            winner = &candidate2;
            loser = &candidate1;
        }

        probvect = update(&probvect, winner, loser, mutrate);

        if isconverged(&probvect, mutrate) {
            break;
        }

        iter = iter + 1;
    }

    let firstresult =  floatstodoubles(&bits2floats(&sample(&probvect)));
    
    let seconresult = hj(fcost, firstresult, 1000, 10.0, 0.0001);
    
    return seconresult;
}

#[cfg(test)]
mod tests {
    use super::mccga;
    use crate::cga::isequal;

    #[test]
    fn test_mccga() {
        fn f(x: &Vec<f64>) -> f64 {
            return (x[0] - 3.14159264).powf(2.0) + (x[1] - 2.718282).powf(2.0);
        }
        let mins: Vec<f64> = vec![-10000.0_f64; 2];
        let maxs: Vec<f64> = vec![10000.0_f64; 2];
        let result = mccga(f, &mins, &maxs, 0.001, 100000);
        assert!(isequal(&result[0], 3.14159265, 0.001));
        assert!(isequal(&result[1], 2.71828, 0.001)); 
    }
}


src/genprobs.rs
use crate::byteworks::*;

use rand::random;

fn getrandomvectorbetween(mins: &Vec<f64>, maxs: &Vec<f64>) -> Vec<f32> {
    let n: usize = mins.len();
    let mut result: Vec<f32> = vec![0.0_f32; n];
    for i in 0..n {
        result[i] = (mins[i] as f32) + random::<f32>() * (maxs[i] as f32 - mins[i] as f32);
    }
    return result;
}

pub fn generate_probability_vector(mins: &Vec<f64>, maxs: &Vec<f64>, ntries: usize) -> Vec<f64> {
    let nbits: usize = mins.len() * 32;
    let mutrate: f64 = (1.0_f64) / (ntries as f64);
    let mut probvector: Vec<f64> = vec![0.0_f64; nbits];
    for _ in 0..ntries {
        let floats: Vec<f32> = getrandomvectorbetween(mins, maxs);
        let floatbits: Vec<u8> = floats2bits(&floats);
        for k in 0..nbits {
            if floatbits[k] == 1 {
                probvector[k] = probvector[k] + mutrate;
            }
        }
    }

    return probvector;
}

#[cfg(test)]
mod tests {
    use super::generate_probability_vector;

    #[test]
    fn test_generateprobabilityvector() {
        let mins: Vec<f64> = vec![0.0_f64, 0.0_f64];
        let maxs: Vec<f64> = vec![10.0_f64, 10.0_f64];
        let prob: Vec<f64> = generate_probability_vector(&mins, &maxs, 10000);
        for i in 0..prob.len() {
            assert!(prob[i] <= 1.0);
            assert!(prob[i] >= 0.0);
        }
    }
}


src/byteworks.rs
pub trait ByteWorks {
    fn to_bitvector(c: f32) -> Vec<u8>;
    fn from_bitvector(v: Vec<u8>) -> f32;
}

impl ByteWorks for f32 {
    fn to_bitvector(c: f32) -> Vec<u8> {
        let u32value: u32 = c.to_bits();
        let mut result: Vec<u8> = vec![0_u8; 32];
        for i in 0..32 {
            result[i] = (((u32value >> i) as u8) & 1) as u8;
        }
        return result;
    }

    fn from_bitvector(v: Vec<u8>) -> f32 {
        let mut u32value: u32 = 0;
        for i in 0..32 {
            u32value = u32value + (v[i] as u32) * (2_u32).pow(i as u32);
        }
        return f32::from_bits(u32value);
    }
}

pub fn floats2bits(v: &Vec<f32>) -> Vec<u8> {
    let mut result: Vec<u8> = Vec::<u8>::with_capacity(v.len() * 32);
    for val in v {
        let mut bitsoffloat = f32::to_bitvector(*val);
        result.append(&mut bitsoffloat);
    }
    return result;
}

pub fn bits2floats(v: &Vec<u8>) -> Vec<f32> {
    let bitsize: usize = v.len();
    let floatssize: usize = bitsize / 32;
    let mut floatvector: Vec<f32> = vec![0.0_f32; floatssize];
    let mut index: usize = 0;
    let mut findex: usize = 0;
    while index + 32 <= bitsize {
        let part: Vec<u8> = v[index..(index + 32)].iter().cloned().collect();
        floatvector[findex] = f32::from_bitvector(part);
        index = index + 32;
        findex += 1;
    }
    return floatvector;
}

pub fn floatstodoubles(x: &Vec<f32>) -> Vec<f64> {
    let n = x.len();
    let mut newvector: Vec<f64> = vec![0.0_f64; n];
    for i in 0..n {
        newvector[i] = x[i] as f64;
    }
    return newvector;
}

#[cfg(test)]
mod tests {
    use super::ByteWorks;
    use super::{bits2floats, floats2bits};

    #[test]
    fn bit_f32_conversation() {
        let val: f32 = 3.14159265_f32;
        let bits: Vec<u8> = f32::to_bitvector(val);
        let otherval: f32 = f32::from_bitvector(bits);
        assert_eq!(val.to_bits(), otherval.to_bits());
        assert_eq!(val, otherval);
    }

    #[test]
    fn test_floats2bits() {
        let v1: Vec<u8> = vec![
            1, 1, 0, 1, 1, 0, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0,
            0, 1, 0, 1, 1, 0, 1, 1, 0, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 1, 0, 0, 0,
            0, 0, 0, 0, 1, 0,
        ];
        let floatvector = vec![3.14159265_f32, 3.14159265_f32];
        let floatsbits: Vec<u8> = floats2bits(&floatvector);
        assert_eq!(floatsbits, v1);
    }

    #[test]
    fn test_bits2floats() {
        let v1: Vec<u8> = vec![
            1, 1, 0, 1, 1, 0, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0,
            0, 1, 0, 1, 1, 0, 1, 1, 0, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 1, 0, 0, 0,
            0, 0, 0, 0, 1, 0,
        ];
        let floats: Vec<f32> = bits2floats(&v1);
        assert_eq!(floats.len(), 2);
        assert_eq!(floats[0], 3.14159265);
        assert_eq!(floats[1], 3.14159265);
    }

    #[test]
    fn test_autotestforconversations() {
        let fvals: Vec<f32> = vec![1.2_f32, 5.6_f32, 78.9998_f32];
        let bits: Vec<u8> = floats2bits(&fvals);
        let floatsback: Vec<f32> = bits2floats(&bits);
        assert_eq!(floatsback.len(), 3);
        assert_eq!(fvals[0], floatsback[0]);
        assert_eq!(fvals[1], floatsback[1]);
        assert_eq!(fvals[2], floatsback[2]);
    }
}


src/hookejeeves.rs

fn clone(x: &Vec<f64>) -> Vec<f64> {
    let n: usize = x.len();
    let mut result: Vec<f64> = vec![0.0_f64; n];
    for i in 0..n {
        result[i] = x[i];
    }
    return result;
}

fn mutate(par: &Vec<f64>, p: usize, d: f64) -> Vec<f64> {
    let mut newpar: Vec<f64> = clone(&par);
    newpar[p - 1] += d;
    return newpar;
}

pub fn hj(
    f: fn(&Vec<f64>) -> f64,
    parv: Vec<f64>,
    maxiter: usize,
    startstep: f64,
    endstep: f64,
) -> Vec<f64> {
    let p = parv.len();
    let mut currentstep: f64 = startstep;
    let mut iter: usize = 0;
    let mut par: Vec<f64> = clone(&parv);
    while iter < maxiter {
        let fold = f(&par);
        let mut fnow: f64 = fold;
        for currentp in 1..=p {
            let mutateleft: Vec<f64> = mutate(&par, currentp, -currentstep);
            let fleft: f64 = f(&mutateleft);
            let mutateright: Vec<f64> = mutate(&par, currentp, currentstep);
            let fright: f64 = f(&mutateright);
            if fleft < fold {
                par = mutateleft;
                fnow = fleft;
            } else if fright < fold {
                par = mutateright;
                fnow = fright;
            }
        }

        if fold <= fnow {
            currentstep = currentstep / 2.0;
        }

        if currentstep < endstep {
            break;
        }

        iter += 1;
    } // end of while

    return par;
}




#[cfg(test)]
mod tests {

    use super::hj;
    use rand;

    #[test]
    fn test_hj_minimization() {
        fn cfmin(x: &Vec<f64>) -> f64 {
            let mut sum: f64 = 0.0;
            for i in 0..x.len() {
                sum += x[i] * x[i];
            }
            return sum;
        }
        let eps = 0.00001;
        let parlen = 20;
        let mut starter = vec![0.0_f64; parlen];
        for i in 1..parlen {
            starter[i] = rand::random();
        }
        let hjresult = hj(cfmin, starter, 10000, 1000.0, 0.000001);
        for i in 0..parlen {
            assert!((hjresult[i] -  0.0_f64).powf(2.0) <= eps);
        }
    }
}

src/lib.rs
pub mod byteworks;
pub mod genprobs;
pub mod cga;
pub mod hookejeeves;


use crate::byteworks::{bits2floats, floatstodoubles};
use crate::cga::{isconverged, sample, update};
use crate::genprobs::generate_probability_vector;
use crate::hookejeeves::hj;

fn binarycost(fcost: fn(&Vec<f64>) -> f64, candidate: &Vec<u8>) -> f64 {
    let doubles = floatstodoubles(&bits2floats(candidate));
    let result = fcost(&doubles);
    return result;
}

pub fn mccga_singleiter(probvect: &mut Vec<f64>, fcost: fn(x: &Vec<f64>)->f64, mutrate: f64) -> (){
    let candidate1: Vec<u8> = sample(&probvect);
    let candidate2: Vec<u8> = sample(&probvect);
    let cost1: f64 = binarycost(fcost, &candidate1);
    let cost2: f64 = binarycost(fcost, &candidate2);
    let mut winner: &Vec<u8> = &candidate1;
    let mut loser: &Vec<u8> = &candidate2;
    if cost2 < cost1 {
        winner = &candidate2;
        loser = &candidate1;
    }
    update(probvect, winner, loser, mutrate);
}

/// Optimize f: R^n -> R type functions in n-dimensional real space.
/// The direction of optimization is supposed to be a minimization.
/// Maximization-typed objective functions can return the negative of the output value.
/// Suppose that the objective function is 
/// ```
/// fn f(x: &Vec<f64>) -> f64 {
///    return (x[0] - 3.14159265).powf(2.0) + (x[1] - 2.71828).powf(2.0);
/// }
/// ```
/// The function is minimized for `x[0] = 3.14159265` and `x[1] = 2.71828`.
/// The function mccga (Machine-coded compact genetic algorithms) is designed to find
/// a solution to that kind of problems.
/// The method initially generates a probability vector using the range of decision variables.
/// Each single element of this probability vector defines the probability of p(b) = 1.0 where
/// b is the corresponding bit. 
/// The algorithm feeds this initial vector to a standard compact algorithm (CGA). In CGA stage
/// of the algorithm, bitwise operations are performed on the IEEE-754 representations of the 
/// 32-bit real values (f32). 
/// In the final stage, the output is fed to a Hooke-Jeeves algorithm for a direct search and 
/// fine-tuning. Finally the algorithm outputs a good approximation for the minimum of the function.
/// Here is the documentation for the details:
/// 
/// Satman, M. H. & Akadal, E. (2020). Machine Coded Compact Genetic Algorithms for Real Parameter Optimization Problems . Alphanumeric Journal , 8 (1) , 43-58 . DOI: 10.17093/alphanumeric.576919
/// 
/// 
/// Here is a example of optimizing the function defined above:
/// ```
/// use mccga::mccga;
/// fn f(x: &Vec<f64>) -> f64 {
///    return (x[0] - 3.14159265).powf(2.0) + (x[1] - 2.71828).powf(2.0);
/// }
/// let mins: Vec<f64> = vec![-10000.0_f64; 2];
/// let maxs: Vec<f64> = vec![10000.0_f64; 2];
/// let result = mccga(f, &mins, &maxs, 0.001, 100000);
/// ```
/// 
pub fn mccga(
    fcost: fn(&Vec<f64>) -> f64,
    mins: &Vec<f64>,
    maxs: &Vec<f64>,
    mutrate: f64,
    maxiter: usize,
) -> Vec<f64> {
    let mut probvect = generate_probability_vector(mins, maxs, 500000);

    let mut iter = 0;

    while iter < maxiter {
        mccga_singleiter(&mut probvect, fcost, mutrate);

        if isconverged(&probvect, mutrate) {
            break;
        }

        iter = iter + 1;
    }

    let firstresult =  floatstodoubles(&bits2floats(&sample(&probvect)));
    
    let seconresult = hj(fcost, firstresult, 1000, 5.0, 0.00001);
    
    return seconresult;
}

#[cfg(test)]
mod tests {
    use super::mccga;
    use crate::cga::isequal;

    #[test]
    fn test_mccga() {
        fn f(x: &Vec<f64>) -> f64 {
            return (x[0] - 3.14159265).powf(2.0) + (x[1] - 2.71828).powf(2.0);
        }
        let mins: Vec<f64> = vec![-10000.0_f64; 2];
        let maxs: Vec<f64> = vec![10000.0_f64; 2];
        let result: Vec<f64> = mccga(f, &mins, &maxs, 0.001, 100000);
        assert!(isequal(&result[0], 3.14159265, 0.0001));
        assert!(isequal(&result[1], 2.71828, 0.0001)); 
    }


    #[test]
    fn test_mccga_5_paramateres() {
        const VALS: [f64; 5]= [7.0, 70.0, 700.0, 7000.0, 70000.0];
        fn f(x: &Vec<f64>) -> f64 {
            let mut sum = 0.0;
            for i in 0..x.len(){
                sum += (x[i] - VALS[i]).powf(2.0);
            }
            return sum;
        }
        let mins: Vec<f64> = vec![-100000.0_f64; 5];
        let maxs: Vec<f64> = vec![100000.0_f64; 5];
        let result: Vec<f64> = mccga(f, &mins, &maxs, 0.001, 100000);
        let delta: f64 = 0.01;
        println!("{:?}", result);
        for i in 0..result.len(){
            println!("Testing {} ?= {}", &result[i], VALS[i]);
            assert!(isequal(&result[i], VALS[i], delta));
        }
    }
}
